Move Table Data Quality to new style

jdunkerley opened the pull request:
### Pull Request Description

- `Column` now returns `ColumnStorage<?>` when `getStorage` is called.
- `Table` now calls `Column.applyMask` rather than accessing the storage.
- `StorageIterators.forEach` now gets a `boolean` value back from the operation, allowing it to exit early without exceptions.
  - Updated `CastOperation` to match this change.
- Moved `CountNothing` into a Unary operation as felt better there!
- `DataQualityMetrics` added to handle all data quality calculations.
  - Only calculated asynchronously.
  - Cached against the storage object.
  - Calculation initialised by the `Column` constructor.
  - Fixes bug with `RequiresNumberFormatting`.
  - Added new `distinct_count` and `count_empty` metrics.
  - Single implementation of sample taking (was one per metric before).
  - Removed all the boiler plate from the Storage classes :)
- Tidied up `IfOperation` to be consistent with other operations.

### Important Notes

<!--
- Mention important elements of the design.
- Mention any notable changes to APIs.
-->

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [ ] The documentation has been updated, if necessary.
- [ ] Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.
- [ ] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      [TypeScript](https://github.com/enso-org/enso/blob/develop/docs/style-guide/typescript.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- [ ] Unit tests have been written where possible.
- [ ] If meaningful changes were made to logic or tests affecting Enso Cloud integration in the libraries, 
      or the Snowflake database integration, a run of the [Extra Tests](https://github.com/enso-org/enso/actions/workflows/extra-nightly-tests.yml) has been scheduled.
  - If applicable, it is suggested to paste a link to a successful run of the Extra Tests.

jdunkerley committed:
Column `getStorage` returns `ColumnStorage<?>`.
Tidy up `IfOperation` a little.

jdunkerley committed:
Java formatting.
Add a boolean abort to the `forEach` iteration.
Altered the CastOperation to use this to short circuit.

jdunkerley committed:
Move CountNothing to dataquality package and restructure to iterators.

jdunkerley committed:
Move the DataQualityMetrics to a new implementation.
If the metric is not calculated then isn't rendered.

jdunkerley committed:
Java format and enso viz changes.

jdunkerley committed:
Docs

GregoryTravis commented on std-bits/table/src/main/java/org/enso/table/data/column/DataQualityMetrics.java:
```diff
@@ -0,0 +1,337 @@
+    return getMetrics(column.getStorage()).getMetrics();
+  }
+
+  public static DataQualityMetrics getMetrics(ColumnStorage<?> columnStorage) {
```
Maybe name these both `get`? Or if not, `getByColumn` and `getByStorage`?

GregoryTravis commented on std-bits/table/src/main/java/org/enso/table/data/column/DataQualityMetrics.java:
```diff
@@ -0,0 +1,337 @@
+  // Default sample size for counting untrimmed cells.
+  public static final long DEFAULT_SAMPLE_SIZE = 10000;
+
+  private static Map<ColumnStorage<?>, DataQualityMetrics> _cachedMetrics;
```
Add a todo to make this an LRU cache, so we don't leak memory; or use a weak reference key.

jdunkerley replied:
It is a LRU Cache: `_cachedMetrics = new LeastRecentlyUsedCache<>(1000);`

jdunkerley committed:
PR comments.

jdunkerley committed:
Use a unique ID to cache the DQ results so that they are not

jdunkerley committed:
Improve data quality metrics.
Alter TableViz to show tool tip based on the quality metric.

jdunkerley committed:
Java format and prettier.

jdunkerley committed:
Remove LongStorageFacade as not used.
Add uniqueID to DoubleStorageFacade.
Add new AnyObject quality indicator with the types.

jdunkerley committed:
Fix ExplodingStorage to meet the new requirements.

jdunkerley committed:
Add message when still computing.

jdunkerley committed:
Fix bug in CountNothing.

jdunkerley committed:
Update lots of changed tests.

jdunkerley committed:
Prettier fails.

jdunkerley committed:
Prettier fails.

jdunkerley committed:
Last bits for column_info.

jdunkerley committed:
Prettier again.
